### PR TITLE
MAINT,ENH: Rewrite scalar math logic

### PR DESCRIPTION
**Pull request by seberg**
This tries to redo the scalar math logic to take some more care about subclasses, but most of all introduce logic to defer to the `other` if `self` can be cast to it safely (i.e. it is the correct promotion).

This makes things much faster and more reliable, since we now use defer much less to `ufuncs` indirectly.  This ensures that integer overflows are reported more reliably.

Another major point about it, is that this reorganizes the coercion of Python int, float, complex (and bool).  This should help a bit with switching to "weak" Python scalars.
    
Further, it contains a commit to make all macros to inline functions and move the floating point overflow flag handling to a return value.  Checking floating point flags is not insanely slow, but it is pretty slow on the scale of integer operations here (~30% on my computer).

It should fix some bugs around subclasses, but then subclassing scalars should still be pretty prone to issues (similar to subclassing arrays I guess).  Complicated subclass cases may still end up in the generic (array) path, but we catch a few more ahead of time.

---

This is the 2-3 approach I have been thinking about.  Another one would be to have (more or less) a single function dealing with any scalar inputs.  It does "inline" the casting logic here.  I do not like that, but it seemed somewhat straight forward – it would be nice to create `npy_cast_@from@_to_@to@` or so functions, to use here more generically.

The alternative would be logic with the existing "cast" functionality, but it should be slower, and while verbose (and the current macros are ugly), I am not sure that would actually end up much better.

The `PyArray_ScalarFromObject` seemed only useful for the old scalar paths, so I added a deprecation:  Neither were even documented, and both would probably need some work to transition to the new DTypes well.
`PyArray_CastScalarDirect` may also be a target, but it is still used in at least one place, so likely should just fix it.

---

Benchmarks:
```
       before           after         ratio
     [732ed25e]       [3b60effc]
     <main>           <scalar-math-rewrite>
-        861±10ns          819±6ns     0.95  bench_scalar.ScalarMath.time_abs('int32')
-         862±3ns          819±5ns     0.95  bench_scalar.ScalarMath.time_abs('float16')
-        929±10ns         882±10ns     0.95  bench_scalar.ScalarMath.time_addition('complex128')
-     1.05±0.02μs          996±4ns     0.95  bench_scalar.ScalarMath.time_addition('float16')
-         859±4ns          813±3ns     0.95  bench_scalar.ScalarMath.time_abs('int16')
-         890±8ns          841±2ns     0.94  bench_scalar.ScalarMath.time_abs('complex64')
-         924±4ns          871±4ns     0.94  bench_scalar.ScalarMath.time_multiplication('float64')
-         957±2ns        895±0.8ns     0.93  bench_scalar.ScalarMath.time_abs('complex256')
-        932±10ns        871±0.4ns     0.93  bench_scalar.ScalarMath.time_addition('float64')
-         922±5ns        855±0.8ns     0.93  bench_scalar.ScalarMath.time_abs('longfloat')
-        893±10ns          826±7ns     0.93  bench_scalar.ScalarMath.time_abs('complex128')
-     1.09±0.07μs          987±8ns     0.90  bench_scalar.ScalarMath.time_multiplication('float16')
-        687±10ns          532±2ns     0.77  bench_scalar.ScalarMath.time_add_int32_other('int32')
-     4.63±0.05μs       3.58±0.1μs     0.77  bench_scalar.ScalarMath.time_addition_pyint('int16')
-     5.58±0.06μs      4.24±0.07μs     0.76  bench_scalar.ScalarMath.time_addition_pyint('float32')
-     5.62±0.08μs       4.22±0.1μs     0.75  bench_scalar.ScalarMath.time_addition_pyint('complex64')
-     1.03±0.01μs          755±9ns     0.73  bench_scalar.ScalarMath.time_add_int32_other('int16')
-     5.76±0.03μs       4.21±0.1μs     0.73  bench_scalar.ScalarMath.time_addition_pyint('float16')
-         922±2ns          662±2ns     0.72  bench_scalar.ScalarMath.time_addition('int16')
-         921±3ns        653±0.5ns     0.71  bench_scalar.ScalarMath.time_multiplication('int64')
-         929±1ns          656±6ns     0.71  bench_scalar.ScalarMath.time_addition('int64')
-         916±8ns          645±4ns     0.70  bench_scalar.ScalarMath.time_multiplication('int16')
-        953±60ns          657±5ns     0.69  bench_scalar.ScalarMath.time_multiplication('int32')
-         934±8ns        642±0.9ns     0.69  bench_scalar.ScalarMath.time_addition('int32')
-     3.54±0.01μs         1.76±0μs     0.50  bench_scalar.ScalarMath.time_power_of_two('longfloat')
-     3.41±0.04μs      1.59±0.01μs     0.47  bench_scalar.ScalarMath.time_power_of_two('float64')
-      3.70±0.1μs         1.65±0μs     0.45  bench_scalar.ScalarMath.time_power_of_two('complex256')
-     2.36±0.05μs         1.04±0μs     0.44  bench_scalar.ScalarMath.time_power_of_two('int64')
-     2.83±0.05μs         1.19±0μs     0.42  bench_scalar.ScalarMath.time_addition_pyint('longfloat')
-     2.87±0.05μs         1.20±0μs     0.42  bench_scalar.ScalarMath.time_addition_pyint('complex256')
-     3.17±0.08μs         1.25±0μs     0.39  bench_scalar.ScalarMath.time_power_of_two('complex128')
-     2.64±0.02μs          992±5ns     0.38  bench_scalar.ScalarMath.time_addition_pyint('complex128')
-     2.17±0.04μs          807±2ns     0.37  bench_scalar.ScalarMath.time_addition_pyint('int64')
-     2.75±0.02μs          993±5ns     0.36  bench_scalar.ScalarMath.time_addition_pyint('float64')
-      14.1±0.2μs      1.24±0.03μs     0.09  bench_scalar.ScalarMath.time_add_int32_other('longfloat')
-      14.6±0.3μs      1.23±0.01μs     0.08  bench_scalar.ScalarMath.time_add_int32_other('complex256')
-      14.1±0.2μs      1.10±0.01μs     0.08  bench_scalar.ScalarMath.time_add_int32_other('float64')
-      14.5±0.1μs      1.11±0.01μs     0.08  bench_scalar.ScalarMath.time_add_int32_other('complex128')
-     13.9±0.07μs          973±6ns     0.07  bench_scalar.ScalarMath.time_add_int32_other('int64')
```
(Some corner cases may be significantly slower, mainly certain `scalar + 0d_array` ops, but I am not sure I want to worry about those much.)

---

**This PR became quite big, I may split it up.  At this time it relies on gh-21178.   It is currently missing some additional tests for the subclass behavior at least, but I would like to check code-coverage on that front as well.**

### CODE REVIEW

**Comment by seberg** [comment]
Hmmmpf, somehow going to `static inline` functions makes clang optimize some floating point errors away for some multiplications.

**Comment by seberg** [comment]
Puh... Added (tricky) tests and fixed some other bugs:
* float16 division was just plain crashing
* small integer division goes to float64 in the ufunc machinery, but went to float32 here.  (It may well have been what division did at some point)

I did not fix that complex comparisons behave differently for arrays and scalars (ufuncs use NaN aware order currently?!).  It is still a change, since the scalar path will be used  a bit more often, but...

Otherwise, these are the scalar-math changes as discussed @mattip.  It is still based off gh-21178, though.

File: numpy/core/src/umath/scalarmath.c.src [hunk 1]
```diff
@@ -275,135 +307,169 @@ static void
...
-            (outp)->imag = (in1i - in1r*rat)*scl;                   \
-        }                                                           \
-    }                                                               \
-    else {                                                          \
-        @rtype@ rat = in2r/in2i;                                    \
-        @rtype@ scl = 1.0@c@/(in2i + in2r*rat);                     \
-        (outp)->real = (in1r*rat + in1i)*scl;                       \
-        (outp)->imag = (in1i*rat - in1r)*scl;                       \
-    }                                                               \
-    } while(0)
```
Comment (seberg) [inline numpy/core/src/umath/scalarmath.c.src:384]: Just to note, since this felt a bit more complex with the branches, I just decided to use the ufunc-loop (and include that) directly.  But happy to go the other way again.

File: numpy/core/tests/test_scalarmath.py [hunk 2]
```diff
@@ -845,3 +878,149 @@ def test_clongdouble_inf_loop(op):
...
+def test_scalar_signed_integer_overflow(dtype):
+    val = np.dtype(dtype).type(8)
+    with pytest.warns(RuntimeWarning, match="overflow encountered"):
+        -val
+
+
+@pytest.mark.parametrize("dtype", np.typecodes["AllInteger"])
+@pytest.mark.parametrize("operation", [
+        lambda val, zero: val // zero,
+        lambda val, zero: val % zero,], ids=["//", "%"])
```
Comment (mattip) [inline numpy/core/tests/test_scalarmath.py:923]: linting
```suggestion
        lambda val, zero: val % zero, ], ids=["//", "%"])
```

**Comment by seberg** [comment]
We should reactivate this, but I am not sure if the test failures were real, so closing/reopening.  (I guess there may ave been something about complex numbers or so.)

**Comment by seberg** [comment]
There are currently two issues.  The first is that on clang+mac, some floating point errors are not reported.  I really don't quite understand why.  Locally on linux (clang 11 though) it doesn't happen (also with `-O3`).  I have tried to use `FINLINE`, since the main change should be the change from macro to function.
(Another try now is adding `volatile` to all of the math `out` functions, but basically, I am out of ideas besides looking at the generated instructions to see what the compiler is doing.  The simple try with godbolt didn't reall help there either.)

The other is that PyPy seems to behave slightly different for `tp_richcompare`, I suspect that is just OK, but should check it.

File: numpy/core/tests/test_scalarmath.py [hunk 3]
```diff
@@ -845,3 +878,152 @@ def test_clongdouble_inf_loop(op):
...
+    ("__pow__", "__rpow__", operator.pow, False),
+    ("__sub__", "__rsub__", operator.sub, False),
+]
+
+
+@pytest.mark.parametrize(["__op__", "__rop__", "op", "cmp"], ops_with_names)
+@pytest.mark.parametrize("sctype", [np.float32, np.float64, np.longdouble])
+@pytest.mark.xfail(IS_PYPY,
+        reason="PyPy does not replace `tp_richcompare` for subclasses so"
+               "our we do not realize when deferring should be preferred.")
```
Comment (seberg) [inline numpy/core/tests/test_scalarmath.py:955]: @mattip I think this PR should be good to go now.  This one is a caveat and I am not sure if I should push it to PyPy.  I think this whole thing is so esoteric that I am not sure I want to worry about it.
The point is that I check `tp_richcompare`, if a subclass has a custom one, we should assume that it knows better what to do than the NumPy provided `tp_richcompare`.  This means, that we should defer.

However, this logic fails on PyPy, because `tp_richcompare` is not replaced (I think) for a subclass overriding e.g. `__le__`.

(The alternative might be to just not try this defering logic at all, but I somewhat feel it should be correct.)

**Comment by seberg** [comment]
Btw. when it comes to the floating point error issues on clang... It turns out the problem was moving the the `fpe_clear_floatstatus_barrier` calls lower.  Somehow using `out` "barrier" seems to have not been enough to ensure the correct order.

Considering that it seemed like it _sometimes_ was correct, I am even wondering if FPEs can be checked out-of-band and it is possible to create race conditions...  But that is more of a curiosity, I am not worried that the fix of moving the FPE checking is not reasonable.

File: numpy/core/src/umath/scalarmath.c.src [hunk 4]
```diff
@@ -1353,36 +1619,52 @@ static NPY_INLINE PyObject *
...
+        case OTHER_IS_SUBCLASS:
+            /*
+             * Success, but defer if not a forward operator.  This implements
+             * the asymmetry of `__add__` and `__radd__` for subclassing.
+             * (A subclass replacing `__add__`/`__radd__` has precedence. It
+             * also means that such a subclass has to always take charge.)
+             */
+            if ((void *)Py_TYPE(other)->tp_richcompare != (void *)@name@_richcompare) {
+                Py_RETURN_NOTIMPLEMENTED;
+            }
```
Comment (mattip) [inline numpy/core/src/umath/scalarmath.c.src:1663]: In both these cases, can you let the `self.tp_richcompare` handle it? That way if a subclass has overridden `tp_richcompare` it should JustWork (or return NotImplemented).

Comment (seberg) [reply numpy/core/src/umath/scalarmath.c.src:1663]: The case that this "fixes" is when you have two subclasses:
```
class myf32(np.float32):
    pass

class my_customf32(np.float32):
    def __eq__(self, other):
        return "something strange"
```
The above logic allows that:
```
myf32(3) == my_customf32(3)
```
defers to `my_customf32`.  The reason is that `myf32` will see that `my_customf32` has custom handling and ask it to do the operation.

If we continue, we return the NumPy behaviour there always (meaning that the operation result will depend on order).  If we return `NotImplemented` always, we would return `False` for the case of:
```python
class myf32_2(np.float32):
    pass

myf32(1) == myf32_2(1)
```
Where two _different_ subclasses which both inherit the comparison are compared.

Comment (seberg) [reply numpy/core/src/umath/scalarmath.c.src:1663]: For most binary operators the rules would be clear, we would defer in `__add__`, but not in `__radd__`.  But comparisons don't provide the op vs. rop information properly :/

Comment (mattip) [reply numpy/core/src/umath/scalarmath.c.src:1663]: > If we continue, we return the NumPy behaviour there always (meaning that the operation result will depend on order). 

This is consistent with Python semantics.

Comment (mattip) [reply numpy/core/src/umath/scalarmath.c.src:1663]: >>> class A(int):
...     def __eq__(self, other):
...         return 'A'
... 
>>> class B(int):
...    pass
... 
>>> B(3) == A(3)
True
>>> A(3) == B(3)
'A'

Comment (seberg) [reply numpy/core/src/umath/scalarmath.c.src:1663]: Arrg, you are right... seems even Python itself doesn't even try to fix this for the other binary operators, the same is true for `__add__` for these builtin classes.

In that case, we can punt and just hope we are fine.  This branch may be more correct, but there is no point in trying harder than Python `int` itself.

Comment (mattip) [reply numpy/core/src/umath/scalarmath.c.src:1663]: This is probably documented somewhere. It is different when the comparison is between instances of a class and a subclass:
```
>>> class A():
...     pass
... 
>>> class B(A):
...    def __eq__(self, other):
...       return 'B'
... 
>>> A() == B()
'B'
>>> B() == A()
'B'
```

Comment (seberg) [reply numpy/core/src/umath/scalarmath.c.src:1663]: Yeah, for direct subclasses Python will ensure the correct order by just calling the subclass first.

There is also this funny assymetry, that if you write Python code at least, you can make `__radd__` mora aggressive than `__add__` to allow a subclass that coerces other subclasses more aggressively to take charge.
But, making that assymetry work in C is very annoying, and it is pretty esoteric anyway.  If even Python `int` bails on it, there is no reason we should not...

Comment (seberg) [reply numpy/core/src/umath/scalarmath.c.src]: OK, removed that additional logic to try to make seperate subclasses work.  I think the remaining "is forward" logic for binops is very similar to what we already have for arrays, and should be fine for PyPy (I cannot think of a failure path, and I don't think it is a change to begin with now).

Comment (mattip) [reply numpy/core/src/umath/scalarmath.c.src]: :100:

**Comment by seberg** [comment]
Ping @mattip, not sure how to proceed.  Do you want to have another look pair programming or can we go ahead?
Or maybe @ganesh-k13 can also have a look, since he had once started looking at similar things.

**Comment by ganesh-k13** [comment]
Thanks for the ping. I'll try to understand this a bit better, was quite lost the first time around :)

File: numpy/core/src/umath/scalarmath.c.src [hunk 5]
```diff
@@ -151,16 +157,16 @@ static void
...
     if (b == 0) {
-        npy_set_floatstatus_divbyzero();
         *out = 0;
+        return NPY_FPE_DIVIDEBYZERO;
     }
 #if @neg@
     else if (b == -1 && a < 0 && a == -a) {
-        npy_set_floatstatus_overflow();
         *out = a / b;
+        return NPY_FPE_OVERFLOW;
```
Comment (mattip) [inline numpy/core/src/umath/scalarmath.c.src:169]: Help me out: when is `a < 0 && a == -a` for bytes?

File: numpy/core/src/umath/scalarmath.c.src [hunk 6]
```diff
@@ -558,196 +630,400 @@ static void
...
+             * always be safe to use the array path.
+             * The issue is, that the other scalar may or may not be designed
+             * to deal with NumPy scalars.  Without knowing that, we cannot
+             * defer (which would be much faster potentially).
+             * TODO: We could add a DType flag to allow opting in to deferring!
+             */
+            ret = OTHER_IS_UNKNOWN_OBJECT;
+    }
+    Py_DECREF(dtype);
+    return ret;
```
Comment (mattip) [inline numpy/core/src/umath/scalarmath.c.src:1018]: I am fine with this big table for now. At some point when we deprecate value based casting for scalars we can rethink this.

File: numpy/core/tests/test_scalarmath.py [hunk 7]
```diff
@@ -66,6 +75,41 @@ def test_leak(self):
...
+        try:
+            res = op(arr1, arr2)
+        except Exception as e:
+            with pytest.raises(type(e)):
+                op(scalar1, scalar2)
+        else:
+            scalar_res = op(scalar1, scalar2)
+            assert_array_equal(scalar_res, res)
+
+
```
Comment (mattip) [inline numpy/core/tests/test_scalarmath.py:112]: Yay for hypothesis-based tests

**Comment by mattip** [comment]
LGTM. The refactoring looks OK, and the speedups are great.

I assume the benchmarks at the top of the PR are still valid?
> Some corner cases may be significantly slower, mainly certain scalar + 0d_array ops ...

Could you add some benchmarks for these and quantify the slowdown?

File: numpy/core/src/umath/scalarmath.c.src (hunk 5 above)
Comment (seberg) [reply numpy/core/src/umath/scalarmath.c.src:169]: Oh, `abs(np.int(-128)) == -128`.

File: numpy/core/src/umath/scalarmath.c.src (hunk 6 above)
Comment (seberg) [reply numpy/core/src/umath/scalarmath.c.src:1018]: Right, created the issue for that!

**Comment by seberg** [comment]
The big slowdown is that for:
```
arr = np.array(1.)
scalar = np.float64(1.)
```
when the array can be cast to the scalar:
```
%timeit scalar + arr
# 287 ns ± 0.303 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```
Used to be much faster (scalar path kicked in), then:
```
%timeit arr + scalar
# 1.67 µs ± 9.22 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```
And for now they both go through the array path.  So this is a huge slowdown in that very specific case.  The `scalar + arr` case there is still 1/3 slower than `scalar + scalar`, since we used to convert the 0-D array to a scalar, and only then try to continue.

I have added a benchmark for it.

**Comment by mattip** [comment]
> So this is a huge slowdown in that very specific case

Sorry, I am not following. What was the new benchmark performance before the PR and what is it afterwards? How does the time for the new benchmarks `time_add_int32arr_and_other` compare to `time_add_int32_other` ?

**Comment by mattip** [comment]
Let's put this in since it does clean up the error handling and makes some parts of the code cleaner. We may need to back it out if it turns out 0d arrays are used more than we think, but I agree scalars are likely to be much more common. I will wait a little while to see if anyone else has an opinion.

**Comment by seberg** [comment]
> We may need to back it out if it turns out 0d arrays are used more than we think, but I agree scalars are likely to be much more common.

Yeah, or re-add special cases for 0-D arrays.  Although, in that case it would be nice to at least also make `arr + scalar` fast and not just `scalar + arr`.  The assymetry adds to my feeling of not wanting to worry much about it.

**Comment by mattip** [comment]
Thanks @seberg
